### DISCUSSION TITLE
Performance comparison. Why is CC65 slower than the others?

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#1395] [General]
https://www.videogamesage.com/topic/762-super-tilt-bro-for-nes/?do=findComment&comment=163145

**Comment by oliverschmidt** [comment]
Other compilers create an IR (https://en.wikipedia.org/wiki/Intermediate_representation) and perform elaborate optimizations on it. In contrast cc65 directly creates assembly code and only performs peephole optimizations (https://en.wikipedia.org/wiki/Peephole_optimization) on it.

**Comment by Fabrizio-Caruso** [comment]
I think GCC-based cross-compilers do so.
I am not sure KickC uses an intermediate representation. 
--
Inviato da myMail per Android domenica, 07 febbraio 2021, 01:52PM +01:00 da Oliver Schmidt  notifications@github.com :

**Comment by mrdudz** [comment]
All the mentioned compilers do, also KickC.

**Comment by Fabrizio-Caruso** [comment]
Thanks. I was wrongly assuming KickC was different than the others as it is only MOS6502 as theIR is also used to support multiple cpu targets.
The fact that they perform so much better makes me think that there may be room for improvement in CC65 even without going through an IR. Maybe by better exploiting zeropage and the hardware stack.
--
Inviato da myMail per Android domenica, 07 febbraio 2021, 02:31PM +01:00 da Bob Andrews  notifications@github.com :

**Comment by mrdudz** [comment]
The reason why they are "so much better" is exactly that

**Comment by Fabrizio-Caruso** [comment]
For example, in the IR they see more registers and some of these virtual registers are mapped into ZP or the hardware stack.
With CC65 you can do sort of the same somehow by mapping global vars to ZP through pragmas. Such optmization might be done in an automatic fashion.
CC65 lacks link-time or any global optimization. How complex would it be?
--
Inviato da myMail per Android domenica, 07 febbraio 2021, 02:39PM +01:00 da Bob Andrews  notifications@github.com :

**Comment by mrdudz** [comment]
uhoh. not that again

**Comment by Fabrizio-Caruso** [comment]
Sorry Bob, I am trying to guess why the IR makes it possible to optimize better. I have read that one of the things it does is to use ZP for virtual registers. I am guessing and I am probably wrong.

**Comment by Fabrizio-Caruso** [comment]
@mrdudz Sorry, I don't want to insist on an old idea I had. I am trying to understand why the IR is so much better (if the tests above are coorect). The IR is making some optimizations possible, which are not present in the current CC65 optimizer. One of these optimizations is the aforementioned virtual registers mapped in ZP. Is there a way to do something similar without the IR? Forget about my previous idea if you think it is bad. I am wondering whether the current architecture can replicate some of the optimizations (not just the ones related to ZP).
What makes the other compilers faster are the optimizations on the IR and not the IR itself. So, maybe, some of these optimizations are possible without IT. Maybe not.

**Comment by mrdudz** [comment]
Those techniques are unrelated

**Reply by Fabrizio-Caruso** [reply]
Great. If virtual registers mapped to ZP is unrelated to IR as you say, then we can have them in CC65. I thought they were closely related to intermediate representation where more registers are possible.

**Comment by mrdudz** [comment]
Sure, we could have that if we wanted it. Someone only would have to rewrite large parts of the code generator :o) I don't think it would gain a lot however, and i actually like that cc65 doesn't use all that much zeropage - that leaves room for assembly that uses them. I don't have performance problems with C code - stuff that really needs to be fast has to be written in assembly anyway on an 8bit platform.

**Reply by Fabrizio-Caruso** [reply]
I understand your point. Ideally this might be done with a parameter to select how much ZP you want to reserve for C. This is somehow already the case in CC65 for register variables (if I remember well). You are right about the limited gain. ZP alone is not going to make miracles. In some of my use-cases it is a 10%-20% improvement at best.
As far as general C performance is concerned, it depends on whether you want to be portable across different architectures and still have decent performance or you want to introduce a lot of Assembly for each of the concerned architectures.

**Reply by Wrathchild70** [reply]
Remember that something like KickC is self-imposing a restriction of non-recursion and so that lends itself to doing stack stuff in a different way to CC65. Many of the apps being written and built with CC65 probably won't employ recursion but that doesn't mean a different optimization strategy can take over. 
Due to KickC doing this, it can analyse the code paths and optimize local variable, method arguments storage. This then lends itself to having multiple labels share the same memory location and debugging gets slightly confusing. When I first tried porting some code to KickC it failed due to exhausting zp quickly and so code-rework was needed to exploit things.
So in that regard even prototyping stuff in CC65 so that it can be debugged more easily and then use KickC to make a more optimized binary is fair enough thing to do.

**Comment by polluks** [comment]
@mrdudz I don't like bad public relations. However videogamesage didn't mention the version of cc65, remember 2.19 doesn't include the whole new optimizations.

**Comment by mrdudz** [comment]
I couldnt care less to be honest :) None of those other compilers is up to the task for what i am using cc65. And for anything that demands performance, Assembly is the weapon of choice anyway. (And i think they actually mentioned somewhere that they were using some recent git version)

**Comment by acqn** [comment]
I've been experimenting a fork with some rough and hackish IR. Maybe for the next major version.

**Comment by oliverschmidt** [comment]
> Maybe for the next major version.

As already discussed it would be great if you would let me/us know when from your perspective the "current" major version is ready.

**Reply by acqn** [reply]
I have serveral potential PRs worth mentioning for the current major version:
- Pointer constness-correct fix/improvement for 'Type': This could make the code clearer.
- More accurate"side-effects tracking (behind the scene)" and "unused code warning (what you see)"
- "Peephole optimizations": They could resolve quite a few open optimation issues but has a cost of large optimizer code addition.
- 
I am OK without only the first one.

**Comment by mrdudz** [comment]
I'm fine with waiting for all of it - but i am also living on the edge and largely ignore the "stable" release :)

**Comment by oliverschmidt** [comment]
> I have serveral potential PRs worth mentioning for the current major version: [...]

Sorry, but even after reading several times I'm not able to understand what you want to express. I need to ask you to elaborate on your thoughts.

**Reply by acqn** [reply]
From my perspective, the "current" major version would be ready with at least the fix for constness-correctness of 'struct Type' pointer usage in code base. I'd like to make other PRs but their absence wouldn't prevent the "current" major release.

Of course, if any serious regression comes up with my recent PRs that have been merged, I wouldn't consider the "current" major version ready.

**Reply by oliverschmidt** [reply]
Thanks for the clarification - now I see :-)

It's sort of a traditiomn that there's roughly one version per year. Because of your contributions the compiler has seen more changes in the last 12 months than in many years ago put together. So I suggest to actually think about tagging a new version when your fix for constness-correctness of 'struct Type' pointer usage in code base is in place - and there are of course no known regressions...
